src/main/java/com/coders/repository/ProductRepository.java

```diff
@@ -33,28 +33,25 @@ public static ProductRepository getInstance() {
     }
 
     public Product addProduct(Product product) throws TheSameNameException, NoDataException, InvalidTypeOfDataException, NotSuchElementException {
-        try {
-            productValidator.validate(product);
-            if (productMap.values().stream().anyMatch(prod -> prod.getName().equals(product.getName()))) {
-                throw new TheSameNameException("Product with the same name already exists: " + product.getName());
-            }
-            int productId = idGenerator.getAndIncrement();
-            product.setId(productId);
-            productMap.put(product.getId(), product);
-            return product;
-        } catch (NoDataException | NotSuchElementException | TheSameNameException | InvalidTypeOfDataException e) {
-            throw e;
-        }
+        productValidator.validate(product);
+        hasProductWithSameName(product);
```

Comment:
tutaj wywołujesz  hasProductWithSameName(product) ale nie sprawdzasz czy ta metoda zwraca true/false i nie rzuczasz wyjatku.
Przez to testAddProductWithTheSameName nie przechodzi